added arduino code for speed setpoint indicator light

baranusluel opened the pull request.

Commit:
added arduino code for speed setpoint indicator light

baranusluel commented:
It seems this commit has appeared under 2Ryan09's username. I'm guessing his username is configured as the default/global git username on the Ubuntu PC in the shop.

2Ryan09 commented:
Great work guys!

A few notes (nothing major):
- For your use of the newTime and oldTime variables, you can reduce your variable count by just having `newTime` be replaced with `millis()` and then setting `oldTime` to `millis()` after calculation of `dps` 
so..
```
newTime = millis();
dps = ((encoder0Pos - encoder0PosPrev) * 360 * 1000 / 800) / (newTime - oldTime);
oldTime = newTime;
```
can be reduced to
```
dps = ((encoder0Pos - encoder0PosPrev) * 360 * 1000 / 800) / (millis() - oldTime);
oldTime = millis();
```
- You can also reduce your variable count by removing `encoder0PosPrev`, just have `encoder0Pos `be reset back to 0 after each loop, else y'all will face an a weird error once y'all get to the max int value (32,767)
- Lastly, the interrupt methods can be reduced a good bit by, instead of comparing the values of each tick to LOW or HIGH, you compare them to each other and adjust the number of ticks that way (refer to MotorShield.ino in the repo for a reference)

JasonGibson274 commented on proj/SpeedLight/SpeedLight.ino:
```diff
@@ -0,0 +1,100 @@
+signed long encoder0PosPrev = 0;
+signed long oldTime = 0;
+signed long newTime = 0;
+signed long dps = 0; // Angulary velocity in degrees per second
```
Angular?

JasonGibson274 commented on proj/SpeedLight/SpeedLight.ino:
```diff
@@ -0,0 +1,100 @@
+ * and compares to arbitrary speed setpoint value. Lights up LED when difference is larger than arbitrary threshold.
+ */
+
+#define encoder0PinA  2
```
I actually really like this as a usage of macros

JasonGibson274 commented on proj/SpeedLight/SpeedLight.ino:
```diff
@@ -0,0 +1,100 @@
+
+void loop() {
+  newTime = millis();
+  dps = ((encoder0Pos - encoder0PosPrev) * 360 * 1000 / 800) / (newTime - oldTime);
```
is 800 a magic number?

JasonGibson274 commented on proj/SpeedLight/SpeedLight.ino:
```diff
@@ -0,0 +1,100 @@
+    }
+  }
+
+  else   // must be a high-to-low edge on channel A
```
Please fix you white spacing here

JasonGibson274 commented on proj/SpeedLight/SpeedLight.ino:
```diff
@@ -0,0 +1,100 @@
+
+  // Look for a high-to-low on channel B
+
+  else {
```
Please fix your white space here

JasonGibson274 reviewed (changes requested):
Overall looks very good to me. Just a couple of questions/nitpciks

baranusluel committed:
Optimized and fixed formatting

baranusluel committed:
Fixed indentation

baranusluel committed:
Fixed indentation (again)

Think something was wrong with my editor, had to fix indentation from Github's online editor

baranusluel committed:
Fixed tabs (for the last time)

Note to self - spaces, not tabs.

baranusluel commented:
Changes have been made

2Ryan09 reviewed (changes requested):
Get into the habit of using ++ and -- when incrementing (in your interrupt methods)

baranusluel committed:
Changed increment/decrement notation

baranusluel commented:
Done. Got too used to Python and MATLAB I guess

JasonGibson274 commented on proj/SpeedLight/SpeedLight.ino:
```diff
@@ -0,0 +1,62 @@
+#define errorThreshold 20
+#define cpr 200 // Cycles per revolution of encoder disk
+
+long targetSpeed = 40; // Target angulary velocity / setpoint in degrees per second
```
Is this supposed to be a long?

baranusluel committed:
Changed long to int

Also removed redundant 'signed' keywords

JasonGibson274 commented on proj/SpeedLight/SpeedLight.ino:
```diff
@@ -0,0 +1,62 @@
+#define errorThreshold 20
+#define cpr 200 // Cycles per revolution of encoder disk
+
+int targetSpeed = 40; // Target angulary velocity / setpoint in degrees per second
```
double?

baranusluel replied:
Didn't think decimal value precision was relevant for angular velocities, especially in this application because error threshold is likely to be significantly larger. Should I change it?

JasonGibson274 replied:
I am not sure what values are you expecting?

2Ryan09 replied:
On that note, I might make a variable in the traditional m/s, then have another variable calculate it in angular velocity to make it easier for anyone coming later. Then make both variables doubles

baranusluel replied:
Considering that the encoder was rotating so fast that the arduino's interrupt frequency couldn't keep up during comp and a new board with an mbed is being designed to fix this, I expect that the speed values will be pretty large. And the error threshold that determines whether an LED warning light should go off will probably be set to a decently large value as well (more than several degrees per second), because if I'm not mistaken this system is meant to indicate when something is very wrong with the robot and human intervention is needed. Therefore I don't think storing the measured and target angular velocities as integers will be a problem.

2Ryan09 replied:
I see where you're coming from. Given the way the code is currently written, an variable of type int would do just fine. However, to have a variable in m/s would be more practical, and since a conversion from m/s to angular would certainly not be clean, both variables would have to be doubles.

JasonGibson274 replied:
we also decreased the encoder to 48 ticks per revolution not 200. in terms of ticks per second you are looking at values of 50 - 114.

baranusluel replied:
Got it. For converting linear target speed to angular, should I define a wheel radius constant?

2Ryan09 replied:
Yes, you can find that constant in MotorShield.ino

baranusluel committed:
Changed for consistency with MotorShield.ino

baranusluel commented:
I've changed integers to floats (same as double for arduino) as discussed. Also copied over constant values from MotorShield.ino, and changed code in general to be more consistent with MotorShield.ino (e.g. using ticks per sec instead of degrees).
